### PR TITLE
Refactor main() to Trainer class

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #948

**What this PR does:**

1. Remove TrainState (This will break the existing checkpoints, sorry)
2. Refactor main() into three pieces:
    1. Trainer.__init__()
    2. Trainer.train()
    3. Trainer.next_batch()
    4. Trainer.train_step()

We keep the original code order (mostly) but only split the code into four methods. We will refactor the internal logic in later diffs.

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -7,22 +7,25 @@
...
 import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.checkpoint.stateful import Stateful
 from torch.distributed.elastic.multiprocessing.errors import record
 
-from torchtitan.components.checkpoint import CheckpointManager, TrainState
-from torchtitan.components.ft import FTParallelDims, init_ft_manager
+import torchtitan.components.ft as ft
+import torchtitan.protocols.train_spec as train_spec_m
```
Comment (tianyu-l) [inline torchtitan/train.py]: What does `_m` stand for? The naming seems not consistent with other things in the repo. Can we find a better explicit naming?

File: torchtitan/train.py [hunk 2]
```diff
@@ -7,22 +7,25 @@
...
+from torchtitan.components.checkpoint import CheckpointManager
 from torchtitan.components.metrics import (
     build_metrics_processor,
     ensure_pp_loss_visible,
 )
 from torchtitan.config_manager import JobConfig
 from torchtitan.distributed import ParallelDims, utils as dist_utils
-
 from torchtitan.protocols.model_converter import build_model_converters
 from torchtitan.protocols.train_spec import get_train_spec
```
Comment (tianyu-l) [inline torchtitan/train.py]: Can we call `get_train_spec` also from the `train_spec` import?

File: torchtitan/train.py [hunk 3]
```diff
@@ -7,22 +7,25 @@
 import os
 import time
 from datetime import timedelta
+from typing import Any, Iterable
 
 import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.checkpoint.stateful import Stateful
```
Comment (tianyu-l) [inline torchtitan/train.py]: Since they appear only once, how about let's not do import and spell out the path inline where they get called? The purpose is to make `import` section "minimal".

File: torchtitan/train.py [hunk 4]
```diff
@@ -31,365 +34,424 @@
...
+    dataloader: train_spec_m.BaseDataLoader
+    metrics_processor: train_spec_m.MetricsProcessor
+    checkpointer: CheckpointManager
+
+    model_parts: list[nn.Module]
+    optimizers: train_spec_m.OptimizersContainer
+    lr_schedulers: train_spec_m.LRSchedulersContainer
+
+    # Enable debug tracing on failure: https://pytorch.org/docs/stable/elastic/errors.html
+    @record
```
Comment (tianyu-l) [inline torchtitan/train.py:57]: Since it's wrapped universally everywhere, can we call / do the wrapping in `main`? Or do we need the fine-grained calling stack info (which we didn't have previously anyway)?

File: torchtitan/train.py [hunk 5]
```diff
@@ -31,365 +34,424 @@
...
+                    torch_profiler.step()
+                if memory_profiler:
+                    memory_profiler.step()
+
+                # reduce timeout after first train step for faster signal
+                # (assuming lazy init and compilation are finished)
+                if self.step == 1:
+                    dist_utils.set_pg_timeouts(
+                        timeout=timedelta(
+                            seconds=self.job_config.comm.train_timeout_seconds
```
Comment (tianyu-l) [inline torchtitan/train.py]:
```suggestion
                            seconds=job_config.comm.train_timeout_seconds
```

File: torchtitan/train.py [hunk 6]
```diff
@@ -31,365 +34,424 @@
...
+
+        self.checkpointer.load(step=job_config.checkpoint.load_step)
+
+        self.train_context = dist_utils.get_train_context(
+            parallel_dims.loss_parallel_enabled,
+            job_config.experimental.enable_compiled_autograd,
+        )
+        self.step = 0
+
+        logger.info("Trainer initialized.")
```
Comment (tianyu-l) [inline torchtitan/train.py:300]: nit
```suggestion
        logger.info("Trainer initialized")
```

File: torchtitan/train.py [hunk 7]
```diff
@@ -31,365 +34,424 @@
...
-            enable_loss_parallel=not job_config.training.disable_loss_parallel,
+class Trainer(Stateful):
+    job_config: JobConfig
+    gc_handler: utils.GarbageCollection
+
+    parallel_dims: ParallelDims
+    train_spec: train_spec_m.TrainSpec
+    world_mesh: dist.DeviceMesh
+
+    ft_manager: ft.FTManager
```
Comment (tianyu-l) [inline torchtitan/train.py]: this field is not used in `train_step` -- I think we can keep it local

File: torchtitan/train.py [hunk 8]
```diff
@@ -31,365 +34,424 @@
...
+            logger.info(f"Running with args: {job_config.to_dict()}")
+
+        # take control of garbage collection to avoid stragglers
+        self.gc_handler = utils.GarbageCollection(gc_freq=job_config.training.gc_freq)
+
+        device_module, device_type = utils.device_module, utils.device_type
+        device = torch.device(f"{device_type}:{int(os.environ['LOCAL_RANK'])}")
+        # Device has to be set before creating TorchFT manager.
+        device_module.set_device(device)
+        self.ft_manager = ft.init_ft_manager(job_config)
```
Comment (tianyu-l) [inline torchtitan/train.py]: ditto

File: torchtitan/train.py [hunk 9]
```diff
@@ -31,365 +34,424 @@
...
+                enable_loss_parallel=not job_config.training.disable_loss_parallel,
+                ft_manager=ft_manager,
+            )
+        dist_utils.init_distributed(job_config)
+
+        # build meshes
+        self.world_mesh = world_mesh = parallel_dims.build_mesh(device_type=device_type)
+        if parallel_dims.dp_enabled:
+            dp_mesh = world_mesh["dp"]
+            self.dp_degree, self.dp_rank = dp_mesh.size(), dp_mesh.get_local_rank()
```
Comment (tianyu-l) [inline torchtitan/train.py]: these seem to be local to `__init__`, too

**Review by tianyu-l — COMMENTED** [review]
looks good in general -- the granularity of changes in this PR is pretty coarse (`init` -> `train` with `next_batch` and `train_step`) so nothing to be worried.

Moving forward, I can see that the interface between output of `next_batch` and `train_step` needs to be more general, e.g. with mask / multimodality needs. But we can do it later.

In this PR, it may be worth setting up a `build_tokenizer_fn` as necessitated by the following use case.
https://github.com/pytorch/torchtitan/pull/919/files#r1980969203

File: torchtitan/train.py (hunk 4 above)
Comment (fegin) [reply torchtitan/train.py:57]: I think we should at least call the two record in `init()` and `train()`. Agree we don't have to call this for the inner functions.

File: torchtitan/train.py (hunk 7 above)
Comment (fegin) [reply torchtitan/train.py]: nice catch!

File: torchtitan/train.py (hunk 3 above)
Comment (fegin) [reply torchtitan/train.py]: That's fine. We can do that since Trainer only inherits from `Stateful`.

File: torchtitan/train.py (hunk 1 above)
Comment (fegin) [reply torchtitan/train.py]: short for `_module`. I changed it back to `_module` since this confuses people.

**Comment by fegin** [comment]
For `build_tokenizer_fn`, I prefer to do it in another PR. So it will be clear if we break something during the refactor.

File: torchtitan/train.py [hunk 10]
```diff
@@ -7,22 +7,23 @@
 import os
 import time
 from datetime import timedelta
+from typing import Any, Iterable
 
 import torch
+import torch.distributed as dist
+import torch.nn as nn
```
Comment (tianyu-l) [inline torchtitan/train.py:14]: It may be too nit, but these two are only called once, too. Can we merge them into call sites using `torch` only?

**Review by tianyu-l — APPROVED** [review]
Looks great! Had one more nit comment.
